tests/arch/arm/arm_sw_vector_relay/src/arm_sw_vector_relay.c

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <linker/linker-defs.h>
+#include <syscall_handler.h>
+
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+
+extern uint32_t _vector_table;
+extern uint32_t __vector_relay_handler;
+extern uint32_t _vector_table_pointer;
+
+uint32_t *vector_table_addr = (uint32_t *)&_vector_table;
+uint32_t *vector_relay_table_addr = (uint32_t *)&__vector_relay_table;
+uint32_t *vector_relay_handler_func = &__vector_relay_handler;
+
+void test_arm_sw_vector_relay(void)
+{
+	/* Verify that the vector relay table entries (except first two
+	 * entries for MSP and ResetHandler) point to the relay handling
+	 * function. Mask with ~0x1 to account for Thumb node.
+	 */
+	uint32_t *vector_relay_table_addr_val =
+		(uint32_t *)((uint32_t)vector_relay_table_addr & ~0x1);
```

Comment:
Well, perhaps the problem lies elsewhere.
https://github.com/zephyrproject-rtos/zephyr/blob/daf303e66034a1e8696118d733d68eaccffb8110/arch/arm/core/aarch32/irq_relay.S#L77
Why is it `GTEXT`, not `GDATA`? This symbol is not a function, is it? @e-rk

Reply:
Yes, it should be exported as `GDATA` instead.

Reply:
Though I have not observed the `0x1` bit being set previously.

Reply:
Not sure if this should be GDATA.
I see, for example that vector_table is exported as GTEXT.
@anangl this is not a function, but the ARM architecture is using it to jump there, during exception entry, and I guess it needs the Thumb mode for this, no? (need to check this)

Reply:
So, I 've done some investigation, and it turns out, it does not matter for the Cortex-m exception entry, whether this address has the Thumb execution flag set or not. The linker, though, as you noticed, adds the thumb flag when the symbol is exported as GTEXT, while it does not when it is passed as GDATA.

I guess, it is safer to keep GTEXT, just in case, we would ever want a soft function to use the relay_vector_table _and_ the real vector_table directly.

Anyhow, this should not be part of this PR; we can open a standalone PR to fix both symbols (relay_vector_table and_vector_table to be exported as GDATA)

Reply:
> I guess, it is safer to keep GTEXT, just in case, we would ever want a soft function to use the relay_vector_table _and_ the real vector_table directly.

Well, I cannot agree with this. To me it's obviously wrong that instead of the address of a given object in memory you get in the application this address with offset 1. The relay vector table (and the real one as well) is an array in memory from which you read something, so you need the exact address of it. It's not a piece of code that gets executed. There is no jump done to this table. The jump is done to the address read from one of its entries. And only in this address the LSB specifies if the target code should be executed in Thumb or ARM state.
In my opinion, both `_relay_vector_table` and `_vector_table` should be exported via `GDATA`. And this should be corrected first, before this PR is merged.

Reply:
> To me it's obviously wrong that instead of the address of a given object in memory you get in the application this address with offset 1. The relay vector table (and the real one as well) is an array in memory from which you read something, so you need the exact address of it.

So far, Zephyr does not use these linker symbols at all, so there's nothing "functionally wrong" here. Also, we are not reading this address in the code, to get anything executed. Even for setting VTOR, we use the `vector_start` symbol, instead. But yeah, by writing this new ARM test, I made use of these symbols - being aware, though, of the fact that the thumb flag is set. :)

> It's not a piece of code that gets executed. There is no jump done to this table. The jump is done to the address read from one of its entries. And only in this address the LSB specifies if the target code should be executed in Thumb or ARM state.

Haven't argued against that; just thought that in the future we might want to use the symbol as a soft function, so it's no need to set this as %object now, and then back to %function :) [since it has worked for us until now, anyways]

>In my opinion, both _relay_vector_table and _vector_table should be exported via GDATA.

That's fine, could you send a PR with the change on on both vector table symbols? Guess this could be done for all ARM variants, but I am also ok if you want to do this for cortex-m only.

> And this should be corrected first, before this PR is merged.
Yeah, that makes sense, since we then would need to adapt the test to remove the masking of the thumb flag.

Reply:
> could you send a PR with the change on on both vector table symbols? Guess this could be done for all ARM variants, but I am also ok if you want to do this for cortex-m only.

Done. See #27039.

Reply:
Thanks for the PR @anangl, this one is now merged. So i'll update the test code to reflect the change.